### PR TITLE
#11972 Fixes bug in BindAuthority zone file loading

### LINKED ISSUE #11972: Twisted Names Error when loading Bind style zone file
Hello Team,

When using twisted names module as `twistd -n dns --bindzone example.com`, we encounter the error:
```
  File "/usr/local/lib/python3.10/dist-packages/twisted/names/authority.py", line 315, in loadFile
    self.origin = nativeString(fp.basename() + b".")
TypeError: can only concatenate str (not "bytes") to str
```
https://github.com/twisted/twisted/blob/c9ae7d730aad4036bb0147f5ce70e6fea46dbb51/src/twisted/names/authority.py#L315

`fb.basename()` returns a string and Python complains about it as the other argument is bytes. 

Thanks.

### PR DESCRIPTION
**Pull request by kipropbrian**
## Scope and purpose

Fixes #11972
This fixes an issues where bind style zone files are not loaded at all, due to an error in how the origin is generated, leading to a concat error. 

`twistd -n dns --bindzone example.com`

`TypeError: can only concatenate str (not "bytes") to str` 

see #11972 for more details

Since nativeString handles both C{bytes} and C{str}, it is now concatenating based on nativeString's output. 

All tests pass locally but I did not write an explicit test to cover this since file I/O in tests is discouraged, and I am assuming that this may be the only way to tests this.

### CODE REVIEW

**Comment by kipropbrian** [comment]
please review

File: src/twisted/names/authority.py [hunk 1]
```diff
@@ -312,7 +312,7 @@ def loadFile(self, filename):
         fp = FilePath(filename)
         # Not the best way to set an origin. It can be set using $ORIGIN
         # though.
-        self.origin = nativeString(fp.basename() + b".")
+        self.origin = nativeString(fp.basename()) + "."
```
Comment (adiroiban) [inline src/twisted/names/authority.py]: Do we still need to use `nativeString` here ?

I think that this was used only for py2 compatibility.

We now support Python 3 only.

**Review by adiroiban — CHANGES_REQUESTED** [review]
Thank for the changes.
Much apprciated.

I did a quick review.
I am not using twisted.names so I don't have much experience here

Can you please also review the automated tests and make sure we have a test that checks this code ?

I see that we already have some tests... I and am wondering why the existing tests haven't detected this bug

https://github.com/twisted/twisted/blob/bb038c658ad8a09273ea8cfcbbb741448367b6dc/src/twisted/names/test/test_names.py#L1209-L1251

File: src/twisted/names/authority.py (hunk 1 above)
Comment (glyph) [reply src/twisted/names/authority.py]: Yep, we should be working to remove any existing uses of `nativeString` and not adding new ones.

**Comment by glyph** [comment]
Ideally we should add a new test or two to cover this bug and ensure it does not regress, but type annotations can also be a big help for bugs like this. If we add annotations to the code and the tests and find that there's a type error — given that there's existing code coverage, I suspect that there is a bug in the relevant test that is just passing the wrong type — we don't necessarily need to add a new test.

Thanks again for your PR @kipropbrian ! If you can address this feedback, that would be great, but if not, please just let us know so we can tag it appropriately and someone else can pick it up.

**Comment by kipropbrian** [comment]
Am jumping on it. Will work on it. Thanks for the feedback.

**Comment by kipropbrian** [comment]
Hey @glyph @adiroiban I have added two test to cover this, one is for loading the bindZone path as a string and the other one as a byte. It seems the path can either be of type byte or string (AnyStr), there is no silent str to byte coercion in Python 3, I suspect this is the root cause. I am new to twisted so I hope the tests are good, I will appreciate any feedback, if you can think of more edge cases I can cover those too. 
The way the previous tests were set up, it fixed the path as type byte in the test set up [#L1224](https://github.com/twisted/twisted/blob/e09d0427fbd980d99c0a6ec6b9a8d329768cc623/src/twisted/names/test/test_names.py#L1224), so there was no case covering Filepaths for the bindzones as type string. 

The issue of nativeString I was not sure how to handle because nativeString seems to be useful even in python3 because FilePath [#L726](https://github.com/twisted/twisted/blob/e09d0427fbd980d99c0a6ec6b9a8d329768cc623/src/twisted/python/filepath.py#L726) and self.path  is of type AnyStr, basename() can return either a byte or a string. So I think nativestring has to handle bytes to string conversion so that we can concat to get `self.origin` [#L315](https://github.com/twisted/twisted/blob/e09d0427fbd980d99c0a6ec6b9a8d329768cc623/src/twisted/names/authority.py#L315), previously with python2 this concatenation was not an issue and nativeString was probably useless, but I may be missing something. 
There is also the limitation to ascii  [#L342](https://github.com/twisted/twisted/blob/e09d0427fbd980d99c0a6ec6b9a8d329768cc623/src/twisted/python/compat.py#L342) which I am not sure how it would affect things if we remove nativeCode and strings are not strictly encoded to ascii but rather unicode, but I did not dive into that. 

I can also work on type annotations but that would need a new PR, should I work on type annotation for the tests first?

I await your feedback. Thanks.

**Comment by kipropbrian** [comment]
please review

File: src/twisted/names/authority.py [hunk 2]
```diff
@@ -302,17 +303,17 @@ class BindAuthority(FileAuthority):
...
     """
 
-    def loadFile(self, filename):
+    def loadFile(self, filename: AnyStr) -> None:
         """
         Load records from C{filename}.
 
         @param filename: file to read from
-        @type filename: L{bytes}
+        @type filename: bytes or str C{AnyStr}
```
Comment (glyph) [inline src/twisted/names/authority.py]: This is not correct use of `AnyStr`.  `AnyStr` is a type variable used to ensure correspondence between bytes/str inputs and bytes/str outputs.  In a context like this, where an input can just be either but it doesn't affect anything else about the signature of the function, it isn't a  It definitely isn't a *third* type that could be returned. In any case, we just don't need `@type` annotations at all anymore, pydoctor supports annotations directly now.
```suggestion
```

File: src/twisted/names/authority.py [hunk 3]
```diff
@@ -302,17 +303,17 @@ class BindAuthority(FileAuthority):
     Supports only C{$ORIGIN} and C{$TTL} directives.
     """
 
-    def loadFile(self, filename):
+    def loadFile(self, filename: AnyStr) -> None:
```
Comment (glyph) [inline src/twisted/names/authority.py]:
```suggestion
    def loadFile(self, filename: bytes | str) -> None:
```

File: src/twisted/names/authority.py (hunk 1 above)
Comment (glyph) [reply src/twisted/names/authority.py]: Given that `nativeString` always returns a `str` now, this was just broken if `filename` was a `bytes` object, which suggests that there's still a missing test if you want the signature to be `bytes | str`.

```suggestion
        self.origin = fp.asTextMode().basename() + "."
```

File: src/twisted/names/test/test_names.py [hunk 4]
```diff
@@ -1221,13 +1223,50 @@ def loadBindString(self, s):
         @return: a new bind authority
         @rtype: L{twisted.names.authority.BindAuthority}
         """
-        fp = FilePath(self.mktemp().encode("ascii"))
+        fp: FilePath[AnyStr] = FilePath(path)
```
Comment (glyph) [inline src/twisted/names/test/test_names.py]: Does mypy complain without this annotation? I would imagine `FilePath[AnyStr]` would be implied by `FilePath(<something: AnyStr>)`. I'd leave it to the inference unless it is required.
```suggestion
        fp = FilePath(path)
```

File: src/twisted/names/test/test_names.py [hunk 5]
```diff
@@ -1221,13 +1223,50 @@ def loadBindString(self, s):
...
         fp.setContent(s)
 
         return authority.BindAuthority(fp.path)
```
Comment (glyph) [inline src/twisted/names/test/test_names.py:1236]: The problem here is that `FileAuthority` doesn't document what the `filename` parameter to its `__init__` is. Either nailing it down to `str` or making `FileAuthority` *itself* `Generic[AnyStr]` would allow this to be addressed more comprehensively.

That isn't required to land this since it might end up touching a bunch more files, but I think we should file it as a follow-up if not.

File: src/twisted/names/test/test_names.py [hunk 6]
```diff
@@ -1221,13 +1223,50 @@ def loadBindString(self, s):
...
 
     def setUp(self):
-        self.auth = self.loadBindString(sampleBindZone)
+        path = self.mktemp().encode("ascii")
```
Comment (glyph) [inline src/twisted/names/test/test_names.py]: Not all filesystems are ASCII, so this may fail randomly.  I think you want `os.fsencode` here if you really need to encode a path to bytes (or maybe just `FilePath(self.mktemp).asBytesMode().path`

File: src/twisted/names/test/test_names.py [hunk 7]
```diff
@@ -1221,13 +1223,50 @@ def loadBindString(self, s):
...
+        )
+
+        self.assertTrue(authority.records, "No records were loaded from the BIND zone")
+
+        self.assertIsInstance(authority.records, dict, "Records is not a dictionary")
+        self.assertGreater(
+            len(authority.records), 0, "No domains were parsed into records"
+        )
+
+    def test_loadBindZonePathAsBytes(self):
```
Comment (glyph) [inline src/twisted/names/test/test_names.py]: New tests should have type annotations so they can be checked.
```suggestion
    def test_loadBindZonePathAsBytes(self) -> None:
```

File: src/twisted/names/test/test_names.py [hunk 8]
```diff
@@ -1318,17 +1357,19 @@ def test_invalidRecordClass(self):
         """
         loadBindString raises NotImplementedError on invalid records.
         """
+        path = self.mktemp().encode("ascii")
```
Comment (glyph) [inline src/twisted/names/test/test_names.py]: Once again, ascii is not necessarily the filesystem encoding.

Comment (glyph) [inline src/twisted/names/test/test_names.py]:
```suggestion
        path = FilePath(self.mktemp()).asBytesMode().path
```

**Review by glyph — CHANGES_REQUESTED** [review]
Thanks for adding the tests, but this has still got a lot of potentially subtle bugs in it, and I'd like to get this fixed in a somewhat more durable way. Right now, it's broken, which means nobody's using it, so obviously it would be nice to get that fixed quickly; but if we fix it halfway and then ship a version that folks *do* start using again, but do it in a way which requires several incompatible breaking changes to the public API to make correct, we will have a bunch of compatibility and upgrade headaches in future versions which a small amount of effort now could avoid.

Thanks for sticking with this and I hope you can do the follow-ups to get it over the finish line! If not, please just let us know if you have to abandon it so someone else can pick it up.

**Comment by kipropbrian** [comment]
Hey @glyph I will definitely follow up on this and get it across the line, thank you so much for the helpful feedback.

File: src/twisted/names/test/test_names.py (hunk 4 above)
Comment (kipropbrian) [reply src/twisted/names/test/test_names.py]: It does not.

File: src/twisted/names/authority.py (hunk 3 above)
Comment (kipropbrian) [reply src/twisted/names/authority.py]: This `bytes | str ` causes tests to fail,  ```builtins.TypeError: unsupported operand type(s) for |: 'type' and 'type'``` seems python3.8 does not support the pipe operator.

Comment (glyph) [reply src/twisted/names/authority.py]: FYI the way we usually fix these is not to use `typing.Union`, since 3.8 support will be dropped pretty soon now. If you add `from __future__ import annotations` to the top of the file, that prevents the annotations from being evaluated at import time (which is faster for newer versions of Python anyway)

**Comment by kipropbrian** [comment]
Please review

File: src/twisted/names/test/test_names.py [hunk 9]
```diff
@@ -1211,7 +1213,7 @@ class BindAuthorityTests(unittest.TestCase):
     Tests for L{twisted.names.authority.BindAuthority}.
     """
 
-    def loadBindString(self, s):
+    def loadBindString(self, s: bytes, path: AnyStr) -> authority.BindAuthority:
```
Comment (adiroiban) [inline src/twisted/names/test/test_names.py]:
```suggestion
    def loadBindString(self, s: bytes, path: AnyStr = '') -> authority.BindAuthority:
```

File: src/twisted/names/test/test_names.py [hunk 10]
```diff
@@ -1221,13 +1223,44 @@ def loadBindString(self, s):
         @return: a new bind authority
         @rtype: L{twisted.names.authority.BindAuthority}
         """
-        fp = FilePath(self.mktemp().encode("ascii"))
+        fp = FilePath(path)
```
Comment (adiroiban) [inline src/twisted/names/test/test_names.py:1233]: this should avoid updating existing test and should reduce the boilerplate code

```suggestion
        if no path:
            path = FilePath(self.mktemp()).asBytesMode().path
        fp = FilePath(path)
```

**Review by adiroiban — APPROVED** [review]
thanks. looks good.

Just a review from someone that has never used Twisted DNS code

I left a minor suggestion that I hope will make it a bit easier to write tests in the future

please consider my suggestion... but it's not blocking

There is also the comment from Glyph about `FileAuthority.filename` ... I don't think that it's required to merge this, but please leave some feedback whether you think it is worth improving that 

Thanks again for the fix.

File: src/twisted/names/test/test_names.py (hunk 9 above)
Comment (kipropbrian) [reply src/twisted/names/test/test_names.py]: Good suggestions, will implement them.

File: src/twisted/names/test/test_names.py (hunk 10 above)
Comment (kipropbrian) [reply src/twisted/names/test/test_names.py:1233]: Thanks for the suggestions and help. I appreciate.

File: src/twisted/names/test/test_names.py (hunk 5 above)
Comment (kipropbrian) [reply src/twisted/names/test/test_names.py:1236]: Once I land this, I will create a new issue to address this.

**Comment by kipropbrian** [comment]
Hey @adiroiban thanks for the suggestions I appreciate your help, I have implemented them. As for @glyph's comments he mentioned FileAuthority's init parameter filename not having a type, I am going to create a new issue to work on it since it may involve more changes as he mentioned, what do you think?

He also mentioned that an extra test may be needed but I was not able to figure out exactly what I should test, since I thought two test one for string and one for byte is enough to cover the issue. Happy to read your thoughts and suggestions.

File: src/twisted/names/test/test_names.py (hunk 9 above)
Comment (kipropbrian) [reply src/twisted/names/test/test_names.py]: Hey @adiroiban setting path to '' or b'' by default is not working. Mypy is complaining and tests are failing. 

```
src/twisted/names/test/test_names.py:1216:55: error: Incompatible default for argument "path" (default has type "str", argument has type "bytes")  [assignment]
src/twisted/names/test/test_names.py:1230:20: error: Incompatible types in assignment (expression has type "bytes", variable has type "str")  [assignment]
```

So I solved this issue a different way. I would be glad to hear your suggestion for changes. I just converted path to be pathType, by default its str but you can make it bytes. Its not the most elegant solution though.

**Comment by kipropbrian** [comment]
Please review

File: src/twisted/names/test/test_names.py [hunk 11]
```diff
@@ -1211,24 +1212,60 @@ class BindAuthorityTests(unittest.TestCase):
...
+    def setUp(self) -> None:
         self.auth = self.loadBindString(sampleBindZone)
 
+    def test_loadBindZonePathAsString(self) -> None:
+        """
+        L{BindAuthority} loads a BIND zone with filepath as type string.
+        """
+        self.assertIsInstance(
+            self.auth, BindAuthority, "Loaded object is not a BindAuthority"
+        )
```
Comment (adiroiban) [inline src/twisted/names/test/test_names.py]: This should make the test self-contained 
```suggestion
        authority = self.loadBindString(sampleBindZone, asText=True)
        self.assertIsInstance(
            authority, BindAuthority, "Loaded object is not a BindAuthority"
        )
```

File: src/twisted/names/test/test_names.py [hunk 12]
```diff
@@ -1211,24 +1212,60 @@ class BindAuthorityTests(unittest.TestCase):
     Tests for L{twisted.names.authority.BindAuthority}.
     """
 
-    def loadBindString(self, s):
+    def loadBindString(
+        self, s: bytes, pathtype: str = "str"
+    ) -> authority.BindAuthority:
```
Comment (adiroiban) [inline src/twisted/names/test/test_names.py]:
```suggestion
    def loadBindString(
        self, s: bytes, asText: bool = True
    ) -> authority.BindAuthority:
```

**Review by adiroiban — APPROVED** [review]
Changes looks good.

I am not happy with the changes in the testing helper... but there aren't many people that care about DNS in Twisted.... so I guess that this is fine to merge.

Check my inline comments and let me know if you would like to apply the changes or merge as is.

thanks again

File: src/twisted/names/test/test_names.py (hunk 11 above)
Comment (kipropbrian) [reply src/twisted/names/test/test_names.py]: This is much better, thanks for the suggestion. Implementing it now.

**Comment by kipropbrian** [comment]
please review

**Comment by adiroiban** [comment]
@glyph do you think we can merge this ? Thanks

**Comment by kipropbrian** [comment]
I would also be interested in following up on any pending issue with a new PR, for example the improvements in FileAuthority's init. I am also wondering if I should squash the commits or will that be handled by github's merge strategy.

File: src/twisted/names/test/test_names.py [hunk 13]
```diff
@@ -1211,23 +1212,58 @@ class BindAuthorityTests(unittest.TestCase):
...
     """
 
-    def loadBindString(self, s):
+    def loadBindString(self, s: bytes, asText: bool = True) -> authority.BindAuthority:
         """
         Create a new L{twisted.names.authority.BindAuthority} from C{s}.
 
         @param s: A string with BIND zone data.
-        @type s: bytes
+        @type s: L{bytes}
```
Comment (glyph) [inline src/twisted/names/test/test_names.py:1220]: FYI, these `@type` and `@rtype` fields are no longer required, as pydoctor can read the annotation directly and will document it properly. It's not worth removing (particularly in a test module) but in the future just avoid adding that redundant info, and remove them anywhere that they match exactly with a type annotation.

**Review by glyph — APPROVED** [review]
